File: app/src/main/java/com/jacobswearingen/fliplauncher/NotificationsViewModel.java

package com.jacobswearingen.fliplauncher;

import android.app.Application;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.service.notification.StatusBarNotification;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class NotificationsViewModel extends AndroidViewModel {
    private final MutableLiveData<List<StatusBarNotification>> notificationsLiveData = new MutableLiveData<>();
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Observer<Void> changesObserver = ignored -> loadNotifications();

    public LiveData<List<StatusBarNotification>> getNotifications() {
        return notificationsLiveData;
    }

    public NotificationsViewModel(@NonNull Application application) {
        super(application);
        NotificationService.getChanges().observeForever(changesObserver);
        loadNotifications();
    }

    @Override
    protected void onCleared() {
        NotificationService.getChanges().removeObserver(changesObserver);
        executor.shutdownNow();
        super.onCleared();
    }

    private void loadNotifications() {
        executor.execute(() -> {
            NotificationManager nm = getApplication().getSystemService(NotificationManager.class);
            List<StatusBarNotification> notifications = NotificationService.getActiveNotificationsList();
            List<StatusBarNotification> sorted = new ArrayList<>(notifications);
            sorted.sort(
                Comparator.comparingInt((StatusBarNotification sbn) -> -getImportance(nm, sbn))
                    .thenComparing(sbn -> Objects.requireNonNullElse(sbn.getNotification().getChannelId(), ""))
                    .thenComparing(Comparator.comparingLong(StatusBarNotification::getPostTime).reversed())
            );
            notificationsLiveData.postValue(sorted);
        });
    }

    private int getImportance(NotificationManager nm, StatusBarNotification sbn) {
        String channelId = sbn.getNotification().getChannelId();
        if (channelId != null && nm != null) {
            NotificationChannel channel = nm.getNotificationChannel(channelId);
            if (channel != null) return channel.getImportance();
        }
        return NotificationManager.IMPORTANCE_DEFAULT;
    }

    public void cancelNotification(String key) {
        NotificationService.cancelNotificationByKey(key);
    }

    public void clearAllNotifications() {
        NotificationService.clearAllSystemNotifications();
    }
}

File: app/src/main/java/com/jacobswearingen/fliplauncher/NotificationsViewModel.kt
package com.jacobswearingen.fliplauncher

import android.app.Application
import android.service.notification.StatusBarNotification
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class NotificationsViewModel(application: Application) : AndroidViewModel(application), NotificationService.Companion.NotificationListener {
    private val _notifications = MutableLiveData<List<StatusBarNotification>>()
    val notifications: LiveData<List<StatusBarNotification>> = _notifications

    init {
        NotificationService.registerListener(this)
        loadNotifications()
    }

    override fun onCleared() {
        super.onCleared()
        NotificationService.unregisterListener(this)
    }

    override fun onNotificationsChanged() {
        loadNotifications()
    }

    @Suppress("DEPRECATION")
    private fun loadNotifications() {
        viewModelScope.launch(Dispatchers.Default) {
            val context = getApplication<Application>()
            val nm = context.getSystemService(android.app.NotificationManager::class.java)
            val sorted = NotificationService.getActiveNotifications()
                .sortedWith(
                    compareByDescending<StatusBarNotification> { it.notification.priority }
                        .thenBy { it.notification.channelId ?: "" }
                        .thenByDescending {
                            val channelId = it.notification.channelId
                            if (channelId != null) nm.getNotificationChannel(channelId)?.importance else null
                                ?: it.notification.priority
                        }
                        .thenByDescending { it.postTime }
                )
            _notifications.postValue(sorted)
        }
    }

    fun cancelNotification(key: String) {
        NotificationService.cancelNotificationByKey(key)
    }

    fun clearAllNotifications() {
        NotificationService.clearAllSystemNotifications()
    }
}